Test/add state use case test

Hasa9nAli opened the pull request:
## Description
This PR implements comprehensive test coverage for the `AddStateUseCase` class, achieving 100% coverage.

## Changes
- Added unit tests for all public and private methods in `AddStateUseCase`
- Covered all possible scenarios including:
  - Successful state addition
  - Invalid project ID case
  - Invalid title validation
  - Project existence check
- Used reflection to test private methods while maintaining encapsulation

## Testing
- [x] Test coverage increased to 100% for `AddStateUseCase`
- [x] Test cases include:
  - `should return true if state add correctly`
  - `should be throw exception if project id is not exist`
  - `should return false when invalid title`
  - `isProjectExist should return false for project not exist`
  - `isValidTitleState should return false for title not valid`

## Notes
- The current implementation of `AddStateUseCase` methods all return false/throw exceptions - actual business logic will be implemented in future PRs
- Used JUnit 5 for testing
- Followed clean test principles with Given/When/Then structure

## Screenshoot 
<img width="499" alt="Screenshot 2025-04-29 at 11 21 11 AM" src="https://github.com/user-attachments/assets/c25f953b-f426-44ae-9904-58824c9adb11" />

Hasa9nAli committed:
add init AddStateUseCase with no implementation

Hasa9nAli committed:
add all test cases of AddStateUseCase

Hasa9nAli committed:
refactor exception assert

ZeyadAbdullah679 commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,102 @@
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import java.util.*
+
```
Formatting should be 4 spaces instead of 1 space

ZeyadAbdullah679 commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,102 @@
+
+class AddStateUseCaseTest{
+ private lateinit var addStateUseCase: AddStateUseCase
+ @BeforeEach
```
You can add a mockk repository and use `every` in each use case to mockk the behaviour of returning values as a list

ZeyadAbdullah679 commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,102 @@
+   val newState = State(
+    id = UUID.randomUUID(),
+    title = "TODO",
+    projectId = UUID.randomUUID()
```
use instead of the random uuid the values you have added inside the every block

ZeyadAbdullah679 commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,102 @@
+ @Test
+ fun `isProjectExist should return false for project not exist`() {
+
+  val method = AddStateUseCase::class.java.getDeclaredMethod("isProjectExist", Project::class.java)
```
Not necessary to do so as testing the invoke function could cover all the code testing branches

AimanYosofiUlt replied:
I think you should test the private function via the public funcation

ZeyadAbdullah679 commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,102 @@
+   )
+
+   // When
+   val result = addStateUseCase.invoke(newState)
```
For short you can use addStateUseCase(newState)

AimanYosofiUlt commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,102 @@
+  }
+
+ @Test
+ fun `should be throw exception if project id is not exist`(){
```
I think you should clarify what the exception will throw here "should be throw IllegalArgumentException exception if project id is not exist"

AimanYosofiUlt commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,102 @@
+  )
+
+  // When && Then
+  val exception = assertThrows<IllegalArgumentException> {
```
I suggest you use a custom exception instead

AimanYosofiUlt commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,102 @@
+  val exception = assertThrows<IllegalArgumentException> {
+   addStateUseCase.invoke(newState)
+  }
+  assertEquals("the project id should be valid", exception.message)
```
I think keep it with one assertion is good practice 

AimanYosofiUlt commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,102 @@
+ }
+
+ @Test
+ fun `should return false when invalid title`() {
```
Sorry bro can you make it more clear message, something like "should return false when adding state with empty title"

AimanYosofiUlt replied:
I think you can make more use-cases for more invalid title names cases

AimanYosofiUlt commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,102 @@
+ @Test
+ fun `isValidTitleState should return false for title not valid`() {
+
+  val method = AddStateUseCase::class.java.getDeclaredMethod("isValidTitleState", String::class.java)
```
Here too:
I think you should test the private function via the public funcation

AimanYosofiUlt reviewed (changes requested):
Hey Hassan I enjoy reviewing your code, but together we can keep improving our work

Hasa9nAli committed:
add custom NoExistProjectException

Hasa9nAli committed:
add isProjectExist method to ProjectsRepository

Hasa9nAli committed:
add projects repository in the constructor and remove unwanted helper method

Hasa9nAli committed:
make
1- space format as 4
2- use evey and verify
3- add project repository to test cases
4- use custom throw exception NoExistProjectException

Hasa9nAli committed:
archive 100% test coverage by call private method

hassanwasfy commented on src/main/kotlin/logic/repositoies/ProjectsRepository.kt:
```diff
@@ -1,5 +1,8 @@
+import java.util.UUID
+
 interface ProjectsRepository {
+    fun isProjectExist(projectId: UUID): Boolean
```
replace from
```kt 
isProjectExist
```
to 
```kt 
isExist
```

it's already defined at the repo name

hassanwasfy replied:
for example you can find the fragment to check if its ready we can use 
```kt
isAdded()
```

@Hasa9nAli

Hasa9nAli committed:
change name of function form isProjectExist to isExist

Hasa9nAli committed:
add addState in repository

Hasa9nAli committed:
rename constructor from repository to projectRepository

Hasa9nAli committed:
fix data type of parameter from string to state

Hasa9nAli committed:
add return false implementation

AimanYosofiUlt reviewed (approved):
Nice work

ZeyadAbdullah679 commented on src/test/kotlin/logic/usecases/AddStateUseCaseTest.kt:
```diff
@@ -0,0 +1,90 @@
+        val validProjectId = UUID.fromString("00000000-0000-0000-0000-000000000001")
+        val state = State(UUID.randomUUID(), "TODO", validProjectId)
+
+        every { projectsRepository.isExist(validProjectId) } returns true
```
You are testing the `isExist` function here whether it returns true or not, so writing for every call of the `isExist` function must return true is wrong. Why? Because I am doing this test to check if the logic behind it is true under certain conditions. What can be done? You can add for every `projectRepository.getAll()` must return a list of projects containing a

ZeyadAbdullah679 reviewed (approved):
Great job 👏
